docs/proposal/20220929-ipam-support.md

```diff
@@ -0,0 +1,423 @@
+- Support dual-stack
+- Allow dynamically allocating and releasing addresses
+- Allow running multiple IPAM providers in parallel
+- Support running any IPAM provider
+
+### Non-Goals/Future Work
+
+- Configuring nameservers from Pool
+- Handling updates to the devices on VSphereMachine objects
+- Handling updates to the fromPools list on individual devices
+- User direct manipulation of IPAddressClaims
+- IP Address Management – this is the job of the IPAM provider
+- Support for specific IPs for specific machines - a machine will receive a
+  random IP in the pool. Implementation of this feature can be left to
+  future work.
+
+## Proposal
+
+Following the contract specified by the IPAM integration proposal we plan to:
+
+- Add the fromPools field to the spec.network.devices section of the
+  VSphereMachine and VSphereMachineTemplate. The fromPools field will allow a
+  user to specify that the network device IPAddress should be claimed from an
+  IPPool.
+- If a VSphereMachine has been specified to claim an IP Address from an IPPool,
+  then CAPV will create an IPAddressClaim, which the IPAM provider will use to
+  allocate an IP Address. This IPAddress can then be used by the
+  VSphereMachine.
+
+### User Stories
+
+#### Story 1
+
+As a user, I want to be able to create a VSphereMachine where the IP Address
+for the machine comes from an IPAM Provider instead of DHCP.
+
+#### Story 2
+
+As a user I want to be able to allocate either IPv4 or IPv6 addresses for my
+VSphereMachine.
+
+#### Story 3
+
+As a user I want to be able to release my IP Address allocation when the
+VSphereMachine is deleted for future VSphereMachines to use.
+
+### Implementation Details/Notes/Constraints
+
+#### API Changes
+
+Adding `fromPools`(type []corev1.TypedLocalObjectReference) to the VSphereMachineTemplate
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: example
+  namespace: vsphere-site1
+spec:
+  template:
+    spec:
+      cloneMode: FullClone
+      numCPUs: 8
+      memoryMiB: 8192
+      diskGiB: 45
+      network:
+        devices:
+        - dhcp4: false
+          dhcp6: false
+          networkName: VM Network
+          fromPools:
+            # reference to the pool
+            - apiGroup: ipam.cluster.x-k8s.io
+              kind: InClusterIPPool
+              name: testpool4
+            - apiGroup: ipam.cluster.x-k8s.io
+              kind: InClusterIPPool
+              name: testpool6
+```
+
+Add `FromPools` to the NetworkDeviceSpec:
+
+```go
+// NetworkDeviceSpec defines the network configuration for a virtual machine's
+// network device.
+type NetworkDeviceSpec struct {
+ // FromPools is a list of references to the pools from which IP addresses should be claimed.
+ FromPools []corev1.TypedLocalObjectReference `json:"fromPools,omitempty"`
+}
+```
+
+#### VSphereVM Controller Changes
+
+1. If the FromPools field is set on a device then this controller will create a

```

Comment:
In this case "on the devices" means "in the device spec of the VSphereMachine object" I think.

Reply:
I agree this is confusing. The device exists on the VM object in the vSphere API before the vm is powered on. The device spec of the VSphereMachine Kubernetes api object describes configuration for that device. 

@geetikabatra I don't quite understand your second question? The IPAM provider will assign one IPAddress per IPAddressClaim. One IPAddressClaim will be created for each entry in the fromPools array. The IPAM provider is configured with the IPs that to be made assignable.

Reply:
Thanks, @schrej, and @tylerschultz. So, I am wondering does the IP Range in `fromPools` is coming from the actual device(NIC) in the device spec of the VSphereMachineObject? There could be many devices like this associated. Would `fromPools`barray change with each VSphereMacgine object?

Reply:
@geetikabatra There is no IP range described in `fromPools`. The `fromPools` section only describes the name, group and kind of a IPAM provider. The IPAM Provider, a separate deployment/controller (not described in depth here), is configured with the pool and ip range. Each VSphereMachine's fromPool entry would receive one IP from the pool's range.

Reply:
Aah! Okay, I get it now. A suggestion, we should not name it to `fromPools`. It should be more intuitive like `IPProvider`

Reply:
I don't think `IPProvider` would be intuitive. The idea of the IPAM contract is to have IP address pools defined in the cluster, which are managed by their corresponding IPAM provider. In theory you could of course have one massive Pool of addresses used by all clusters,  but it's also likely that each cluster uses their own pool.
Since you interact with Pools, and not the provider, when configuring where an Address comes from, `pool` should be included in the name.
Maybe something like `addressesFromPools` would be more clear.

FWIW: For the metal3 provider it's [going to be called](https://github.com/metal3-io/cluster-api-provider-metal3/pull/769) `fromPoolRef`, but there you can only reference a single pool as the configuration looks a bit different in general.

Reply:
`addressesFromPools` also sounds good to me.

Reply:
Ok, sounds like a plan. We'll rename to `addressesFromPools`.